// tsc-esm-fix/target/cjs/index.cjs
"use strict";
const {
  __export,
  __reExport,
  __toCommonJS
} = require('./esblib.cjs');


// src/main/ts/index.ts
var ts_exports = {};
__export(ts_exports, {
  DEFAULT_FIX_OPTIONS: () => import_options.DEFAULT_FIX_OPTIONS,
  fix: () => import_fix.fix,
  fixBlankFiles: () => import_legacy.fixBlankFiles,
  fixContents: () => import_legacy.fixContents,
  fixDefaultExport: () => import_legacy.fixDefaultExport,
  fixDirnameVar: () => import_legacy.fixDirnameVar,
  fixFilenameExtensions: () => import_fix.fixFilenameExtensions,
  fixFilenameVar: () => import_legacy.fixFilenameVar,
  fixModuleReferences: () => import_legacy.fixModuleReferences,
  fixSourceMapRef: () => import_legacy.fixSourceMapRef
});
module.exports = __toCommonJS(ts_exports);
var import_fix = require("./fix.cjs");
var import_options = require("./options.cjs");
var import_legacy = require("./legacy.cjs");
__reExport(ts_exports, require("./interface.cjs"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_FIX_OPTIONS,
  fix,
  fixBlankFiles,
  fixContents,
  fixDefaultExport,
  fixDirnameVar,
  fixFilenameExtensions,
  fixFilenameVar,
  fixModuleReferences,
  fixSourceMapRef,
  ...require("./interface.cjs")
});
