// tsc-esm-fix/target/cjs/options.cjs
"use strict";
const {
  __spreadValues,
  __spreadProps,
  __export,
  __toCommonJS
} = require('./esblib.cjs');


// src/main/ts/options.ts
var options_exports = {};
__export(options_exports, {
  DEFAULT_FIX_OPTIONS: () => DEFAULT_FIX_OPTIONS,
  JS_EXT: () => JS_EXT,
  TS_EXT: () => TS_EXT,
  normalizeOptions: () => normalizeOptions
});
module.exports = __toCommonJS(options_exports);
var import_util = require("./util.cjs");
var JS_EXT = [".js", ".jsx", ".mjs", ".mjsx", ".cjs", ".cjsx"];
var TS_EXT = [".ts", ".tsx", ".mts", ".mtsx", ".cts", ".ctsx"];
var DEFAULT_FIX_OPTIONS = {
  cwd: process.cwd(),
  tsconfig: ["./tsconfig.json"],
  src: [],
  target: [],
  filenameVar: true,
  dirnameVar: true,
  ext: true,
  tsExt: TS_EXT,
  jsExt: JS_EXT,
  unlink: true,
  debug: () => {
  }
  // eslint-disable-line
};
var normalizeOptions = (opts = {}) => __spreadProps(__spreadValues(__spreadValues({}, DEFAULT_FIX_OPTIONS), (0, import_util.omitUndefinedKeys)(opts)), {
  tsconfig: opts.tsconfig ? (0, import_util.asArray)(opts.tsconfig) : DEFAULT_FIX_OPTIONS.tsconfig,
  src: (0, import_util.asArray)(opts.src),
  target: (0, import_util.asArray)(opts.target),
  jsExt: typeof opts.jsExt === "string" ? opts.jsExt.split(",") : opts.jsExt || DEFAULT_FIX_OPTIONS.jsExt,
  tsExt: typeof opts.tsExt === "string" ? opts.tsExt.split(",") : opts.tsExt || DEFAULT_FIX_OPTIONS.tsExt,
  debug: typeof opts.debug === "function" ? opts.debug : opts.debug === true ? console.log : DEFAULT_FIX_OPTIONS.debug
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_FIX_OPTIONS,
  JS_EXT,
  TS_EXT,
  normalizeOptions
});
