// tsc-esm-fix/target/esm/fixes/index.mjs
import { fixBlankFiles } from './fix-blank-files.mjs';
import { fixDefaultExport } from './fix-default-export.mjs';
import { fixDirnameVar, fixFilenameVar } from './fix-dirname-var.mjs';
import { fixModuleReferences } from './fix-module-ref.mjs';
import { fixSourceMapRef } from './fix-sourcemap-ref.mjs';
export { fixBlankFiles } from './fix-blank-files.mjs';
export { fixFilenameVar, fixDirnameVar } from './fix-dirname-var.mjs';
export { fixModuleReferences } from './fix-module-ref.mjs';
export { fixSourceMapRef } from './fix-sourcemap-ref.mjs';
export { fixDefaultExport } from './fix-default-export.mjs';
export const fixContents = (ctx) => {
    const { options } = ctx;
    let _ctx = ctx;
    if (options.ext)
        _ctx = fixModuleReferences(_ctx);
    if (options.dirnameVar)
        _ctx = fixDirnameVar(_ctx);
    if (options.filenameVar)
        _ctx = fixFilenameVar(_ctx);
    if (options.fillBlank)
        _ctx = fixBlankFiles(_ctx);
    if (options.forceDefaultExport)
        _ctx = fixDefaultExport(_ctx);
    if (options.sourceMap)
        _ctx = fixSourceMapRef(_ctx);
    return _ctx;
};
